va/va_prot.h

```diff
@@ -0,0 +1,364 @@
+ *
+ * \code
+ * VAEntrypoint *entrypoints;
+ * int i, num_entrypoints, supportsProtectedDisplayLink = 0;
+ *
+ * num_entrypoints = vaMaxNumEntrypoints();
+ * entrypoints = malloc(num_entrypoints * sizeof(entrypoints[0]);
+ * vaQueryConfigEntrypoints(va_dpy, VAProfileProtected, entrypoints, &num_entrypoints);
+ *
+ * for (i = 0; !supportsProtectedDisplayLink && i < num_entrypoints; i++) {
+ *     if (entrypoints[i] == VAEntrypointProtectedDisplayLink)
+ *         supportsProtectedDisplayLink = 1;
+ * }
+ * \endcode
+ *
+ * Then, the vaGetConfigAttributes() function is used to query the protection capabilities.
+ *
+ * \code
+ * VAConfigAttrib attrib_prot[2];
+ * VAConfigAttribValProtectedDisplayLink wired;
+ * VAConfigAttribValProtectedDisplayLink wireless
+ * attrib_prot[0].type = VAConfigAttribDisplayLinkHdcpWired;
+ * attrib_prot[1].type = VAConfigAttribDisplayLinkHdcpWireless;
+ * vaGetConfigAttributes(va_dpy, VAProfileProtected, VAEntrypointProtectedDisplayLink, attrib_prot, 2);
+ * wired = attrib_prot[0].value;
+ * VAConfigAttribDisplayLinkHdcpWireless = attrib_prot[0].value;
+ * if ((wired.bits.session_type & VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0) == 0) {
+ *     // not find desired mode
+ *     assert(0);
+ * }
+ * \endcode
+ *
+ * \section api_prot_setup Set up a protected session
+ *
+ * A protected session is a session between application and GPU hardware.
+ * This protected session can be attached to the display. Also with protected session,
+ * application can do specific TEE execution functionality with TEE inside GPU, but it depends
+ * on IHV's implementation.
+ * Before creating a protected session, it needs to create a config first via
+ * vaCreateConfig(). Then using this config id to create a protected session via
+ * vaCreateProtectedSession().
+ *
+ * The general control flow is demonstrated by the following pseudo-code:
+ * \code
+ * // Create config
+ * VAConfigID config_id;
+ *
+ * wired.bits.session_type = VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0;
+ * wired.bits.protocol_version = VA_PROTECTED_DISPLAY_LINK_HDCP_VERSION_22;
+ * attrib_prot[0].value = wired.value;
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected, VAEntrypointProtectedDisplayLink, attrib_prot, 1, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected session via vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID protected_session;
+ *
+ * va_status = vaCreateProtectedSession(
+ *     va_dpy,
+ *     config_id,
+ *     &protected_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_prot_attach Attach/Detach protected session to the display which want to enable/disable protection
+ *
+ * Protected session is attached to display to enable protection.
+ *
+ * \code
+ * vaAttachProtectedSession(va_dpy, 0, protected_session);
```

Comment:
Is it possible use use `vaCreateContext/vaDestroyContext` instead of `vaCreateProtectedSession/vaDestroyProtectedSession`? So we can hide `vaAttachProtectedSession/vaDetachProtectedSession` details in `vaBeginPicture(va_dpy, protected_session, surface)/vaEndPicture(va_dpy, protected_session)`?

Could we use `vaGetDisplayAttributes/vaSetDisplayAttributes` instead of `vaProtectedSessionHwUpdate/vaProtectedSessionExecute`? You can pass the protected id as param.

Reply:
Hi @xuguangxin, 

Thanks for your feedback.

The first question is about using vaCreateContext/vaDestroyContext instead of vaCreateProtectedSession/vaDestroyProtectedSession. At the beginning, I was planning to using vaCreateContext to create protected session, but I found that there are functions with parameter VAContextID but can't be used by protected session. For example, vaBeginPicture/vaRenderPicture/vaEndPicture are not applied on protected session. That's why I create VAProtectedSessionID and a new sets of API for it.

Second question is embedded vaAttachProtectedSession/vaDetachProtectedSession in vaBeginPicture/vaEndPicture. I proposed these functions is to link protected session with decode/encode/vp context. They can attached at frame level or whole video level depends on the situation. For example, in some DRM playback, there are few seconds clear frame at the beginning, then starting encrypted frame. At this case, protected session are not attached at the beginning and start to attached when encrypted frames are coming.

The third question is using vaGetDisplayAttributes/vaSetDisplayAttributes instead of vaProtectedSessionHwUpdate/vaProtectedSessionExecute. I can use them but I do no think they are same concept. For vaProtectedSessionExecute, it's doing some functions inside TEE or GPU, not just get attributes. Also we need to define the  VAConfigAttribType if we use vaGetDisplayAttributes. I'm not going to define function id, will leave it as IHV to define them.

Hope these will answer your questions, thanks.

Reply:
> The first question is about using vaCreateContext/vaDestroyContext instead of vaCreateProtectedSession/vaDestroyProtectedSession.

Actually, I don't consider "I found that there are functions with parameter VAContextID but can't be used by protected session" to be a valid argument. Yes, there are functions which can't be used. In this case, driver will need to check context type and error out if wrong function was called.

Each time you define new functions you give additional burden to support them. And vaCreateContext has some arguments which you don't have in your new API, but might find them useful in the future. I really suggest to use old vaCreate/DestroyContext functions...

Reply:
With creating VAProtectedSessionID instead of using VAContextID, there are few benefits of it.
1. For vaAttachProtectedSession, it's a different concept that using VAProtectedSessionID will show that we are going to attach a protected session to the existing VA context (decode/encode/vp). If using VAContextID, it may confuse application developer that if the 3rd argument can be used existing VA context or not.
2. We have few newly defined API that will only be used with VAProtectedSessionID. If using VAContextID, it may confuse application developer that if these API can be used with existing VA Context or not.

Reply:
> vaBeginPicture/vaRenderPicture/vaEndPicture are not applied on protected session. That's why I create VAProtectedSessionID and a new sets of API for it

driver attached some data to decode_ctx, can you put your protected session id there?

>At this case, protected session are not attached at the beginning and start to attached when encrypted frames are coming.

can we use VABuffer to indicate it's clear or encrypted? so we can pass it to driver using VaRenderPicture.

>The third question is using vaGetDisplayAttributes/vaSetDisplayAttributes instead of vaProtectedSessionHwUpdate/vaProtectedSessionExecute

could you explain more about the function, it's frame level, or stream level? sync or async? How to use it? Why libva's user need care about it. 

thanks

Reply:
#1 Although this approach may seem easy fix but has significant short comings -
1.a Please note protected session concept is not subset of decode(why? please see prologue and background on HWDRM) but if we force protected session to be part of decode then design has following shortcomings - 
1.a.(i) Protected session needs to be applied for encode or video processing so proposed secure primitives from protected session will not be limited to decode but has to be either maintained for encode and video processing or we have to create decode context for encode or video processing which does not make sense since we are creating artificial dependencies while LibVA does not create these dependencies - Say if we maintain secure primitive across all 3(decode/encode/video processing) then this will cause duplication that will result in all sorts of not so easy to debug issues. 
1.a.(ii) Say proposed secure primitive is applied to all 3 then next problem will arise when decode/encode/video processing has to operate on different protected sessions(yes, it is common - Example - when secure decoding is done by one key and encoding is done by another key) then we may end-up creating very complex design. 
1.a.(iii) Say proposed secure primitive is applied to all 3 that has taken care of multiple sessions then next problem will arise when decode/encode/video processing pipelines are processing mix of clear and protected frame(yes, it is very common and being productized by Google WV) and so on(encryption formats, sub-sample/full-sample)... List will continue since decoding is not designed(and rightfully so) for keeping protected session in perspective... this is the reason CENC spec is not part of H.264 or H.265 or VP9 etc.  
1.b Protected sessions has to perform TEE related tasks(please see previous discussion about provisioning and license acquisition) so if we make protected session part of decode then decode will have to do TEE tasks which is not media primitives. 
#2 We could if concept of clear/encrypted is binary BUT Alas! concept of encrypted is not straight forward as clear since encrypted bit-streams or frames could be encrypted using multiple schemes. For example, according to CENC spec, encrypted samples could be full-sample or sub-sample encrypted besides this sub-sample encryption could have stripping of clear/encrypted sample... Point is construct of encryption is versatile depending on if it is being applied to bit-stream or decoded outputs or video processed surfaces.
#3 This goes back to the question of - How does HWDRM playback SW/TEE/HW pipeline looks like? I have previously replied to this question and added pipeline diagram and sequence diagram. Bottom line is, session creation in TEE is inherently different from using this TEE session to perform certain TEE tasks such as license acquisition or provisioning etc.